### ISSUE TITLE
Add support for a unique page visit ID

### ISSUE THREAD

**Issue by englehardt** [citp/OpenWPM#26]
Right now the database is keyed by `crawl_id` and `top_url`. If we want to visit the same url multiple times with the same crawler we will have multiple visits in the database with the same effective key. Instead, we should develop a notion of a `visit_id` which keys all data for a specific site visit.

It's possible to implement this in the following way

1. Add a `page_visits` table to the database with schema
```
create table page_visits {
    id integer primary key,
    url VARCHAR(500) NOT NULL,
}
```
2.  When `TaskManager` initializes it pulls `max(id)` from `page_visits` and sets a class property `max_visit_id`.

3. For each `get` (or `browse`) command issued, `TaskManager`:
    * assigns `max_visit_id` to a `last_visit_id` property of the `Browser` object corresponding to the browser the `get`/`browse` command was issued to.
    * saves a new record in the `page_visits` table corresponding to the new `id` and `url`
    * increments `max_visit_id`

4. Any new commands issued to that `Browser` will use the `last_visit_id` property to key all data generated by that command. Since the `Browser` object exists in the main thread, this ID will need to explicitly be passed with each command.

**Comment by englehardt** [comment]
I see some benefit to keeping `top_url` around in many cases, but switch it to be the current url in the address bar. The url requested will still be saved in the `site_visits` table. There are several use cases where this is helpful, e.g. determining if an HTTP request is first for third-party for a page that redirects to a new domain on visit. 

In the current model, let's say we visit `A.com` and it immediately redirects to `B.com` which then makes a request to `B.com/file.js`. Here `http_requests.top_url = A.com` and `http_requests.url = B.com/file.js`, so comparing hostnames will falsely label the request as a third-party.

**Comment by dreisman** [comment]
@englehardt : Right now in #74 we've just done away with top_url without replacing it with the current URL from the address bar like you suggested in your last comment. Is it even possible to get the URL in the address bar from the proxy? Or is this yet another reason to switch back to logging from a firefox extension?

At least #74 isn't a regression from the current state regarding top_urls.

**Comment by englehardt** [comment]
> Is it even possible to get the URL in the address bar from the proxy? Or is this yet another reason to switch back to logging from a firefox extension?

I don't see a way to do this reliably. There's no way to do it within the proxy. We could have selenium grab the current URL and pass it to the proxy, but what we pass depends on timing. If we grab the current url immediately after loading the page, the page may be halfway through a redirect chain so we'd still be labeling things with the incorrect URL.

I think extension-based instrumentation is the only way to get this. As far as I can tell, the extension should always have access to the current top-level URL and will be able to grab updated versions as it changes.

**Comment by englehardt** [comment]
Fixed in #74
